### PR TITLE
graphql-ws: optimise the "resolve" routine

### PR DESCRIPTION
**Pull request by oliver-sanders**
* Partially address
* The resolve routine is implmented recursively in the graphql-ws library.
* Because the function is async this results in a large number of async tasks being created when the library is used with large, deeply nested schema.
* Async tasks have an overhead, above that of regular function calls.
* For the example in https://github.com/cylc/cylc-uiserver/issues/547 this resulted in over 10 seconds of overheads.

Results **96% saving!**:
* Before (`base_async.resolve`):
  * cumtime: 11.24s
  * calls: 268'053
* After: (`resolve.resolve`):
  * cumtime - 0.3965s
  * calls: 24

Suggest adding this routine to the cylc-uiserver library for now as the graphql-ws library is not currently maintained (see #333).

The original implementation can be found here: https://github.com/graphql-python/graphql-ws/blob/7ef25ecfaf5390bf1a0cb4023272d8cb074368c2/graphql_ws/base_async.py#L37-L62

TODO:
* [x] Reviewers, please replicate profiling results.
* [x] Changelog entry.
* [x] License legal stuff.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

File: cylc/uiserver/websockets/resolve.py
Comment (wxtim) [inline cylc/uiserver/websockets/resolve.py:30]: Not what?

File: cylc/uiserver/websockets/resolve.py
Comment (wxtim) [inline cylc/uiserver/websockets/resolve.py:42]: I thought that de-recursivizing was the whole point (tm)

**Comment by wxtim** [comment]
IMO Adding it and modifying it : Would be nice if these were separate commits to keep that info in the repo history?

**Comment by wxtim** [comment]
foo2

File: cylc/uiserver/websockets/resolve.py
Comment (wxtim) [inline cylc/uiserver/websockets/resolve.py:66]: - Spaces make it clear that there are three different if loops.
- English notes.
- `items` => `children`

```suggestion
        # Await actual data if _data is awaitable:
        if is_awaitable(_data):
            _data = await _data
            if isinstance(_data, Promise):
                _data = _data.value
            if _container is not None:
                _container[_key] = _data

        # Populate child tasks from data:
        if isinstance(_data, dict):
            children = _data.items()
        elif isinstance(_data, list):
            children = enumerate(_data)
        else:
            children = None

        # If there are child tasks, put them into the stack:
        if items is not None:
            stack.extend([
                (child, _data, key)
                for key, child in items
            ])
```

**Review by wxtim — COMMENTED** [review]
- [x] Read code. I have views on readability, but I'm happy that the change replicates the functioning of the original. Might be nice to unit test it?
- [x] Tried it out. Looks like there is a considerable speed gain for larger workflows.

Not pressed the approve button because it's still draft, presumably indicating futher work to come?

Might we unit test this function?

File: cylc/uiserver/websockets/resolve.py
Comment (oliver-sanders) [reply cylc/uiserver/websockets/resolve.py:66]: Agreed, but, defo out of scope, but moreover this kinda isn't our code...

This file is a temporary measure pending the merge of a PR in an upstream repo which is held up by #333, so we should try to keep changes here minimal (and push changes to the upstream repo when it's ready).

File: cylc/uiserver/websockets/resolve.py
Comment (oliver-sanders) [reply cylc/uiserver/websockets/resolve.py:30]:
```suggestion
# NOTE: transient dependency from graphql-ws purposefully not
# reflected in cylc-uiserver dependencies 
```

File: cylc/uiserver/websockets/resolve.py
Comment (oliver-sanders) [reply cylc/uiserver/websockets/resolve.py:42]:
```suggestion
    Wait on any awaitable children of a data element and resolve
```

File: cylc/uiserver/websockets/resolve.py
Comment (oliver-sanders) [inline cylc/uiserver/websockets/resolve.py:28]: @dpmatthews, can you check you're happy with this.

File: cylc/uiserver/websockets/resolve.py
Comment (wxtim) [inline cylc/uiserver/websockets/resolve.py:31]:
```suggestion
# reflected in cylc-uiserver dependencies
```

**Review by wxtim — APPROVED** [review]
Took the liberty of removing the trailing whitespace causeing the style check to fail.

Do you want to appoint a second review?

**Comment by oliver-sanders** [comment]
The actual diff is trivial and we're choked up so might just slide this through on 1.

**Pull request merged**
